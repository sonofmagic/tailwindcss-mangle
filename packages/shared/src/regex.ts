export function escapeStringRegexp(str: string) {
  if (typeof str !== 'string') {
    throw new TypeError('Expected a string')
  }
  return str.replaceAll(/[$()*+.?[\\\]^{|}]/g, '\\$&').replaceAll('-', '\\x2d')
}

export function makeRegex(
  str: string,
  options: {
    exact: boolean
  } = {
    exact: true,
  },
) {
  return new RegExp(`(?<=^|[\\s"])${escapeStringRegexp(str)}${options.exact ? '(?=$|[\\s"])' : ''}`, 'g')
}
