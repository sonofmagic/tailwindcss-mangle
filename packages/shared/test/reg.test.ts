import { makeRegex } from '@/regex'

describe('regex', () => {
  it('z-10 regex', () => {
    const testCase = 'z-10 w-full max-w-5xl items-center justify-between font-mono text-sm lg:flex'
    const regex = makeRegex('z-10')
    const arr = [...testCase.matchAll(regex)]
    expect(arr.length).toBe(1)
    expect(arr).toMatchSnapshot()
  })

  it('trailing slash should exact match case 0', () => {
    const testCase = 'bg-red-500 bg-red-500/50'
    const regex = makeRegex('bg-red-500', {
      exact: false,
    })
    const arr = [...testCase.matchAll(regex)]
    expect(arr.length).toBe(2)
    expect(arr).toMatchSnapshot()
  })

  it('trailing slash should exact match case 1', () => {
    const testCase = 'bg-red-500 bg-red-500/50'
    const regex = makeRegex('bg-red-500')
    const arr = [...testCase.matchAll(regex)]
    expect(arr.length).toBe(1)
    expect(arr).toMatchSnapshot()
  })

  it('trailing slash should exact match case 2', () => {
    const testCase = 's bg-red-500 bg-red-500/50 bg-red-500'
    const regex = makeRegex('bg-red-500')
    const arr = [...testCase.matchAll(regex)]
    expect(arr.length).toBe(2)
    expect(arr).toMatchSnapshot()
  })
})
