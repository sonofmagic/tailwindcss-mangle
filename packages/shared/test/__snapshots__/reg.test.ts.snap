// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`regex > trailing slash should exact match case 0 1`] = `
[
  [
    "bg-red-500",
  ],
  [
    "bg-red-500",
  ],
]
`;

exports[`regex > trailing slash should exact match case 1 1`] = `
[
  [
    "bg-red-500",
  ],
]
`;

exports[`regex > trailing slash should exact match case 2 1`] = `
[
  [
    "bg-red-500",
  ],
  [
    "bg-red-500",
  ],
]
`;

exports[`regex > z-10 regex 1`] = `
[
  [
    "z-10",
  ],
]
`;
